internal/gcs/guestconnection.go

```diff
@@ -0,0 +1,129 @@
+package gcs
+
+import (
+	"context"
+	"fmt"
+	"net"
+	"sync"
+
+	"github.com/Microsoft/go-winio"
+	"github.com/Microsoft/hcsshim/internal/schema1"
+	"github.com/sirupsen/logrus"
+
+	"github.com/Microsoft/go-winio/pkg/guid"
+)
+
+const (
+	protocolVersion = 4
+
+	gcsVsockPort            = 0x40000000
+	firstIoChannelVsockPort = 0x40000001
+
+	nullContainerID = "00000000-0000-0000-0000-000000000000"
+)
+
+type GuestConnection struct {
+	listener  net.Listener
+	brdg      *bridge
+	vmID      guid.GUID
+	mu        sync.Mutex
+	nextPort  uint32
+	notifyChs map[string]chan struct{}
+	caps      schema1.GuestDefinedCapabilities
+	log       *logrus.Entry
+}
+
+// NewGuestConnection creates a new GCS connection for the given VM.
+func NewGuestConnection(vmID guid.GUID, log *logrus.Entry) (*GuestConnection, error) {
+	l, err := winio.ListenHvsock(&winio.HvsockAddr{
+		VMID:      vmID,
+		ServiceID: winio.VsockServiceID(gcsVsockPort),
+	})
+	if err != nil {
+		return nil, err
+	}
+	gc := &GuestConnection{
+		listener:  l,
+		vmID:      vmID,
+		nextPort:  firstIoChannelVsockPort,
+		notifyChs: make(map[string]chan struct{}),
+		log:       log,
+	}
+	return gc, nil
+}
+
+// Capabilities returns the guest's declared capabilities.
+func (gc *GuestConnection) Capabilities() *schema1.GuestDefinedCapabilities {
+	return &gc.caps
+}
+
+// Protocol returns the protocol version that is in use.
+func (gc *GuestConnection) Protocol() uint32 {
+	return protocolVersion
+}
+
+// Connect starts the guest connection.
+func (gc *GuestConnection) Connect(ctx context.Context) error {
+	conn, err := gc.listener.Accept()
+	gc.listener.Close()
+	if err != nil {
+		return err
+	}
+	gc.brdg = newBridge(conn, gc.notify, gc.log)
+	gc.brdg.Start()
+	req := negotiateProtocolRequest{
+		MinimumVersion: protocolVersion,
+		MaximumVersion: protocolVersion,
+	}
+	var resp negotiateProtocolResponse
+	resp.Capabilities.GuestDefinedCapabilities = &gc.caps
+	err = gc.brdg.RPC(ctx, rpcNegotiateProtocol, &req, &resp, true)
+	if err != nil {
+		return err
+	}
+	if resp.Version != protocolVersion {
+		return fmt.Errorf("unexpected version %d returned", resp.Version)
+	}
+	return nil
+}
+
+// GuestRequest sends a custom guest request to the GCS.
+func (gc *GuestConnection) GuestRequest(ctx context.Context, settings interface{}) error {
```

Comment:
I'm not sure I understand the need for this. It seems like it should just be replaced with `ModifySettings` since it is targeting the UVM and does not allow any other ID.

Reply:
Then sends a ModifySettings request. Are you just saying it should be renamed to ModifySettings?

Reply:
I can rename it to Modify for consistency with Container and HCS, but it's not quite the same...

Reply:
I agree its not the same but it just seemed that the HCS called this `ModifySettings` and if its targeted at the `UVM ID` (`nullContainerID`) then it means its a "host" modification. `GuestRequest` here seems to me like `Container` should call this method to issue a `GuestRequest`